### PR TITLE
framework/st_things : fix memory leakage to some extent when repeating easysetup

### PR DESCRIPTION
**Pull request by sushant-23**
signed off by: sushant-23 sushant.g@samsung.com

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [f01af566556765b2566699e4547bd9e927f69876] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [f01af566556765b2566699e4547bd9e927f69876] - Code Rule Check (C++) OK.

**Comment by sunghan-chang** [comment]
@sushant-23 @uvasistha-samsung @harish-191 As @JeongJunSik left comment on #1967 , to contribute on TizenRT, you guys must follow our rules. Please find our rules on our wiki.

**Comment by Hyunjun85** [comment]
please change title like as "framework/tizenrt: [S18TRST-9]Fixes memory leakage to some extent when repeating easy setup".

**Comment by sunghan-chang** [comment]
Dont write [s18xxx] stubs at commit title. Even though I am a Samsung engineer, I cant understand what it is. Here is a public repository, not Samsung internal.

**Comment by sushant-23** [comment]
@Hyunjun85 yes I have applied only your patch and it gave less memory leak. I am looking for further improvement.

**Comment by seinfra** [comment]
Target : [bdda92c5705307db2fcccda5045ac0bf472c35be] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [bdda92c5705307db2fcccda5045ac0bf472c35be] - Code Rule Check OK.

File: framework/src/st_things/things_stack/framework/things_security_manager.c [hunk 1]
```diff
@@ -798,6 +798,8 @@ const unsigned char g_regional_test_root_ca[] = {
  *
  * NOTE : This API should be invoked after sm_generate_mac_based_device_id invoked.
  */
```
Comment (JeongJunSik) [inline framework/src/st_things/things_stack/framework/things_security_manager.c]: This comment is misplaced
Delete it or put it in correct place.

File: framework/src/st_things/things_stack/framework/things_security_manager.c [hunk 2]
```diff
@@ -798,6 +798,8 @@ const unsigned char g_regional_test_root_ca[] = {
...
+static g_b_init_key = false;
```
Comment (JeongJunSik) [inline framework/src/st_things/things_stack/framework/things_security_manager.c]: you miss data type.

**Comment by seinfra** [comment]
Target : [09620b3941fcc3ae9d6bc3918c5f5886dd2b1924] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [09620b3941fcc3ae9d6bc3918c5f5886dd2b1924] - Code Rule Check (C++) OK.

**Comment by heuisam** [comment]
I think this commit should be divided into three commits as below. Becase these three things cause memory leaks, but each has its different cause.
And the 2nd is temporary solution, you have to change sequence of this function as I sent by email.
The things_sss_key_handler_init and things_sss_rootca_handler_init move to the other function because  save_signed_asymmetric_key is called several times by iotivity stack and it makes dangling pointer.
 
-  wrong type of g_certChain
-  overwrite pointer of devicecert.data without free in things_sss_key_handler_init
-  missing things_log_shutdown in deinit

**Comment by heuisam** [comment]
hmm, I expected you guys to get the right solution for 2nd one when I sent my patch to you guys by email.
You guys just changes author.

**Comment by sushant-23** [comment]
hello @heuisam , @Hyunjun85 , what should I do next? 
And I had one more query: In the function `make_cloud_address` in `framework/cloud_manager.c`,
`ipport` is assigned ip_port which is later concated to `g_cloud_address`. Hence we can free ipport as it is not used again. But freeing it gives the following error during easy setup:  
`T18 ERROR [things_ping] is_ip_key_equal`.
Can you please help me with this.

File: framework/src/st_things/things_stack/framework/things_security_manager.c [hunk 3]
```diff
@@ -829,15 +832,17 @@ static OCStackResult save_signed_asymmetric_key(OicUuid_t *subject_uuid)
 
 #ifdef USE_SSS
 	if (dm_get_easy_setup_use_artik_crt()) {
-		if (things_sss_key_handler_init() < 0) {
+		if (!g_b_init_key && things_sss_key_handler_init() < 0) {
```
Comment (Taejun-Kwon) [inline framework/src/st_things/things_stack/framework/things_security_manager.c]: This is error log, log should include information about g_b_init_key 's state & return value of things_sss_key_handler_init()

File: framework/src/st_things/things_stack/framework/things_security_manager.c [hunk 4]
```diff
@@ -829,15 +832,17 @@ static OCStackResult save_signed_asymmetric_key(OicUuid_t *subject_uuid)
...
 			THINGS_LOG_E(TAG, "InitializeSSSKeyHandlers() Fail");
 			return OC_STACK_ERROR;
 		}
+		g_b_init_key = true;
 
-		if (things_sss_rootca_handler_init(subject_uuid) < 0) {
+		if (!g_b_init_cert && things_sss_rootca_handler_init(subject_uuid) < 0) {
```
Comment (Taejun-Kwon) [inline framework/src/st_things/things_stack/framework/things_security_manager.c]: same as above.

File: framework/src/st_things/things_stack/framework/things_security_manager.c (hunk 3 above)
Comment (sushant-23) [reply framework/src/st_things/things_stack/framework/things_security_manager.c]: hi @Taejun-Kwon , should I put this in logs `THINGS_LOG_E(TAG, "g_b_init_key = 0 and InitializeSSSKeyHandlers() Fail");`  and  same thing for g_b_init-cert ?

**Comment by seinfra** [comment]
Target : [02a59c72011b1cd36857119593d40de1fd44b082] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [02a59c72011b1cd36857119593d40de1fd44b082] - Code Rule Check (C++) OK.

**Comment by Hyunjun85** [comment]
@SenthilKumarGS Could you please review this PR.
If It is unnecessary PR, Coudle please close this.

**Comment by Hyunjun85** [comment]
I have closed this PR.
This is not validation to us.

**Pull request closed without merging**